frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/GitFetchSupportImpl.java

```diff
@@ -0,0 +1,463 @@
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import com.intellij.dvcs.MultiMessage;
+import com.intellij.dvcs.MultiRootMessage;
+import com.intellij.internal.statistic.IdeActivity;
+import com.intellij.notification.NotificationType;
+import com.intellij.openapi.progress.EmptyProgressIndicator;
+import com.intellij.openapi.progress.ProcessCanceledException;
+import com.intellij.openapi.progress.ProgressManager;
+import com.intellij.openapi.progress.util.BackgroundTaskUtil;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.vcs.VcsException;
+import com.intellij.openapi.vcs.VcsNotifier;
+import com.intellij.util.concurrency.AppExecutorUtil;
+import git4idea.GitUtil;
+import git4idea.commands.Git;
+import git4idea.commands.GitAuthenticationGate;
+import git4idea.commands.GitAuthenticationListener;
+import git4idea.commands.GitImpl;
+import git4idea.commands.GitRestrictingAuthenticationGate;
+import git4idea.config.GitConfigUtil;
+import git4idea.fetch.GitFetchResult;
+import git4idea.fetch.GitFetchSupport;
+import git4idea.repo.GitRemote;
+import git4idea.repo.GitRepository;
+import io.vavr.Lazy;
+import io.vavr.Tuple;
+import io.vavr.Tuple2;
+import io.vavr.collection.HashMap;
+import io.vavr.collection.List;
+import io.vavr.collection.Map;
+import io.vavr.control.Option;
+import io.vavr.control.Try;
+import kotlin.jvm.functions.Function1;
+import lombok.AllArgsConstructor;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import com.virtuslab.gitmachete.frontend.actions.toolbar.FetchAllRemotesAction;
+import com.virtuslab.logger.IPrefixedLambdaLogger;
+import com.virtuslab.logger.PrefixedLambdaLoggerFactory;
+
+/**
+ * @deprecated This implementation is a workaround to provide features missing in pre-2020.1 versions.
+ * The main benefit of it is specific branch update ({@link GitFetchSupportImpl#fetch(GitRepository, GitRemote, String)}).
+ * We need this class for {@link FetchAllRemotesAction} and {@link BasePullBranchAction}.
+ */
+@Deprecated
+public final class GitFetchSupportImpl implements GitFetchSupport {
+
+  private static final IPrefixedLambdaLogger LOG = PrefixedLambdaLoggerFactory.getLogger("frontendActions");
+
+  private final Project project;
+  private final ProgressManager progressManager = ProgressManager.getInstance();
+
+  private final AtomicInteger fetchRequestCounter = new AtomicInteger(0);
+
+  private GitFetchSupportImpl(Project project) {
+    this.project = project;
+  }
+
+  private static final AtomicReference<@Nullable GitFetchSupportImpl> INSTANCE = new AtomicReference<>(null);
+
+  public static GitFetchSupportImpl fetchSupport(Project project) {
+    var gitFetchSupport = INSTANCE.get();
```

Comment:
Actually AtomicReference isn't enough here :/ if two `fetchSupport` invocations run concurrently, they can still end up returning different values... in fact I'd suppress the checkstyle against `synchronized` here (`@SuppressWarnings("regexp")` IIRC) ;)

Reply:
I'd even say that `AtomicReference` isn't needed if `fetchSupport` is `static synchronized`

Reply:
(Assuming that `instance` is only ever accessed via `fetchSupport`)

Reply:
Or instead! consider one of the `AtomicReference` more complex methods... `updateAndGet` maybe?

Reply:
Actually, while you're here: almost identical comments apply to `com.virtuslab.gitcore.impl.jgit.GitCoreBranch#getPointedCommit`

Reply:
Or instead! https://www.vavr.io/vavr-docs/#_lazy, but TBD if that's thread-safe (I suspect so, given that most of Vavr is a calque of Scala constructs, and Scala's `lazy val` is thread-safe by design)

Reply:
> Actually, while you're here: almost identical comments apply to `com.virtuslab.gitcore.impl.jgit.GitCoreBranch#getPointedCommit`

`getPointedCommit` throws :cry:

Reply:
:cry: ok, then `synchronized` for `getPointedCommit` maybe...

Reply:
done